File: 121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut res = 0;
        let mut minSoFar = prices[0];
        for price in prices {
            res = res.max(price - minSoFar);
            minSoFar = minSoFar.min(price);
        }
        res
    }
}

File: code/leetcode/121-best-time-to-buy-and-sell-stock/121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int res = 0, minSoFar = prices[0];
        for (auto price: prices) {
            res = max(price - minSoFar, res);
            minSoFar = min(minSoFar, price);
        }
        return res;
    }
};